src/base/init/init.c

```diff
@@ -362,6 +369,10 @@ void low_mem_init(void)
     leavedos(98);
   }
 
+  phys_low = roundUpToNextPowerOfTwo(LOWMEM_SIZE + EXTMEM_SIZE + XMS_SIZE);
+  memsize = phys_low;
+  if (config.dpmi)
+    memsize += config.dpmi_base + HUGE_PAGE_ALIGN(dpmi_mem_size());
```

Comment:
I guess LOWMEM_SIZE is accounted twice
here. Is that intentional?
I.e. it it in both dpmi_base and phys_low.

Reply:
yes it's intentional to align, so that the bottom 21 bits of both phys and linear address are both 0x110000.

so I have this patch now vs your mapmem branch.

```
diff --git a/src/base/init/init.c b/src/base/init/init.c
index 0cd9fcb89..d97912825 100644
--- a/src/base/init/init.c
+++ b/src/base/init/init.c
@@ -372,8 +372,7 @@ void low_mem_init(void)
   phys_low = roundUpToNextPowerOfTwo(LOWMEM_SIZE + EXTMEM_SIZE + XMS_SIZE);
   memsize = phys_low;
   if (config.dpmi)
-    memsize += config.dpmi_base + HUGE_PAGE_ALIGN(dpmi_mem_size()) -
-        LOWMEM_SIZE;
+    memsize += config.dpmi_base + HUGE_PAGE_ALIGN(dpmi_mem_size());
   mem_base = mem_reserve(memsize);
   mem_base_mask = ~(uintptr_t)0;
 #ifdef __x86_64__
@@ -401,11 +400,11 @@ void low_mem_init(void)
   ptr += phys_low;
   phys_rsv = phys_low - (LOWMEM_SIZE + HMASIZE);
   /* create non-identity mapping up to phys_low */
-  ptr2 = smalloc_aligned_topdown(&main_pool, NULL, PAGE_SIZE, phys_rsv);
+  ptr2 = smalloc_fixed(&main_pool, MEM_BASE32(memsize - phys_rsv), phys_rsv);
   assert(ptr2);
   if (config.dpmi) {
     void *dptr = smalloc_fixed(&main_pool, MEM_BASE32(config.dpmi_base),
-        dpmi_mem_size());
+        ptr2 - MEM_BASE32(config.dpmi_base));
     assert(dptr);
     if (config.cpu_vm_dpmi == CPUVM_KVM) {
       /* map dpmi+uncommitted space to kvm */
```
to  a) the alignment b) have no hole in the two smalloc'ed areas, and b) with smalloc_fixed no need for #1949 if $_dpmi=(0).

Of course this could be cleaned up a bit more (e.g. merging the two smalloc's since they are adjacent). I could also move the first smalloc below dpmi_base, then it's naturally aligned at the right place, but we'd need an extra mmap_kvm.

So either some extra work now or in a future PR?

Reply:
Thanks.
Please see the new patch.
I added most of the above (in some
form at least), but I still don't understand
why `ptr2 - MEM_BASE32(config.dpmi_base)`.
Is it somehow "broken" with a hole?

Reply:
Yes it's broken otherwise because that hole isn't mapped in kvm but can still be allocated by dpmi 0504. It's just there for alignment.

I'll have a look at the patch later.

Reply:
Oh no sorry the kvm map is big enough, it's just two areas of linear allocation then. So this is probably ok for now. I'll do a few tests later and will let you know.

Reply:
OK but now I added the check to
avoid that (will squash later).
Is that enough?

Reply:
I was more thinking of this:
```
@@ -453,7 +453,9 @@ dpmi_pm_block * DPMI_mallocLinear(dpmi_pm_block_root *root,
        return NULL;
 
     if (base == -1)
-       realbase = smalloc_aligned_topdown(&main_pool, NULL, PAGE_SIZE, size);
+       realbase = smalloc_aligned_topdown(&main_pool,
+               &dpmi_lin_rsv_base[dpmi_lin_mem_rsv()],
+               PAGE_SIZE, size);
     else
        realbase = smalloc_fixed(&main_pool, MEM_BASE32(base), size);
     if (realbase == NULL) {
```
but then `dpmi_lin_mem_free()` is still off.

Also I think you need this:
```
@@ -200,7 +200,7 @@ int dpmi_lin_mem_rsv(void)
 {
     if (!config.dpmi)
        return 0;
-    return PAGE_ALIGN(config.dpmi - (LOWMEM_SIZE + HMASIZE));
+    return PAGE_ALIGN(config.dpmi_base - (LOWMEM_SIZE + HMASIZE));
 }
 
 int dpmi_lin_mem_free(void)
```

But this all starts to be a bit too complicated for not much benefit, so here's my present patch:
https://gist.github.com/3385a7087d6c64af286a63d9cc78d7d1
which maps xms/extmem below dpmi_base instead of above it. This way all the linear maps (dpmi 0x0504, LFB, xms/extmem) are between 0x110000 and dpmi_base.

Sure this creates an extra kvm map slot, but this is correctly aligned and only temporary, since later I can make it identity mapped, once the physical extmem/xms maps to kvm_base.

Reply:
Good idea, added these 2 hunks.
As for new patch - I don't see why
it is any better to map below dpmi_base.
Extra fragmentation, and for what
benefit?

Reply:
Should now be fixed.

Reply:
needs a tiny change for $_dpmi=(0), 

```
diff --git a/src/base/init/init.c b/src/base/init/init.c
index e0082e29a..25c622478 100644
--- a/src/base/init/init.c
+++ b/src/base/init/init.c
@@ -401,12 +401,13 @@ void low_mem_init(void)
   ptr += phys_low;
   phys_rsv = phys_low - (LOWMEM_SIZE + HMASIZE);
   /* create non-identity mapping up to phys_low */
-  ptr2 = smalloc_topdown(&main_pool, phys_low);
+  ptr2 = smalloc_topdown(&main_pool, config.dpmi ? phys_low : phys_rsv);
   assert(ptr2);
-  /* unused hole for alignment */
-  ptr2 += LOWMEM_SIZE + HMASIZE;
   if (config.dpmi) {
-    void *dptr = smalloc_fixed(&main_pool, MEM_BASE32(config.dpmi_base),
+    void *dptr;
+    /* unused hole for alignment */
+    ptr2 += LOWMEM_SIZE + HMASIZE;
+    dptr = smalloc_fixed(&main_pool, MEM_BASE32(config.dpmi_base),
         dpmi_mem_size());
     assert(dptr);
     if (config.cpu_vm_dpmi == CPUVM_KVM) {
```

otherwise passes the below tests, so that's good :)

```
test/test_dos.py --require-attr=cputest PPDOSGITTestCase
test/test_dos.py PPDOSGITTestCase.test_memory_dpmi_japheth
test/test_dos.py PPDOSGITTestCase.test_memory_xms
test/test_dos.py PPDOSGITTestCase.test_memory_emm286_borland
```